src/codecs/png.rs

```diff
@@ -158,83 +160,165 @@ impl<R: BufRead + Seek> PngDecoder<R> {
+        // Check the iTXt chunk.
         if let Some(mut itx_chunk) = self
             .reader
             .info()
             .utf8_text
             .iter()
-            .find(|chunk| chunk.keyword.contains(XMP_KEY))
+            .find(|chunk| keys.iter().any(|key| chunk.keyword.contains(key)))
             .cloned()
         {
             itx_chunk.decompress_text().map_err(ImageError::from_png)?;
             return itx_chunk
                 .get_text()
-                .map(|text| Some(text.as_bytes().to_vec()))
+                .map(|text| Some(text.into_bytes()))
                 .map_err(ImageError::from_png);
         }
-        Ok(None)
-    }
 
-    fn iptc_metadata(&mut self) -> ImageResult<Option<Vec<u8>>> {
+        // Check the zTXt chunk. This chunk is special as the metadata stored in it for all common forms is
+        // expected to be stored as a hex string, which we decode manually here.
         if let Some(mut text_chunk) = self
             .reader
             .info()
             .compressed_latin1_text
             .iter()
-            .find(|chunk| IPTC_KEYS.iter().any(|key| chunk.keyword.contains(key)))
+            .find(|chunk| keys.iter().any(|key| chunk.keyword.contains(key)))
             .cloned()
         {
             text_chunk.decompress_text().map_err(ImageError::from_png)?;
             return text_chunk
                 .get_text()
-                .map(|text| Some(text.as_bytes().to_vec()))
+                .map(|text| Some(parse_raw_profile(text.as_ref(), header)))
                 .map_err(ImageError::from_png);
         }
 
+        // Check the TEXt chunk.
         if let Some(text_chunk) = self
             .reader
             .info()
             .uncompressed_latin1_text
             .iter()
-            .find(|chunk| IPTC_KEYS.iter().any(|key| chunk.keyword.contains(key)))
+            .find(|chunk| keys.iter().any(|key| chunk.keyword.contains(key)))
             .cloned()
         {
-            return Ok(Some(text_chunk.text.into_bytes()));
+            return Ok(Some(text_chunk.text.as_bytes().to_vec()));
         }
+
         Ok(None)
     }
+}
+
+fn unsupported_color(ect: ExtendedColorType) -> ImageError {
+    ImageError::Unsupported(UnsupportedError::from_format_and_kind(
+        ImageFormat::Png.into(),
+        UnsupportedErrorKind::Color(ect),
+    ))
+}
+
+/// Decodes a hex-encoded string with skipped headers and interspersed newlines.
+/// This is how ImageMagick stores metadata in zTXt chunks. By default, metadata
+/// there looks like the following:
+///
+/// "Raw profile type exif\nOptional\n264223\nExif\x00\x004578..."
+///  ^^^^^^^^^^^^^^^^^^^^   ^^^^^^^   ^^^^^^  ^^^^^^^^^^^ ^^^^
+///      Metadata key    Optional info  Size  (opt.) header  payload (hex encoded)
+///
+/// This method parses such data and returns the decoded payload.
+/// If the structure doesn't match, we return and empty `Vec<u8>`.
+fn parse_raw_profile(buffer: &str, header: Option<&str>) -> Vec<u8> {
```

Comment:
Is this something that would make sense to expose as a public method from the `png` crate?